### PR TITLE
BloomShipper: add cache for downloaded blocks

### PR DESCRIPTION
**Pull request by vlad-diachenko**
adapted embeddedcache.go to store downloaded bloom blocks on local filesystem.

**What this PR does / why we need it**:
This cache will be used by bloom-gateway to not download the blocks each time. 

In the cache we store the objects(`cachedBlock` struct) that contain: 
1. `blockDirectory`: the `path` to the directory where the block was extracted 
2. `activeQueriers`: thread-safe counter of active block users. This field is important because we do not want the directory to be removed while it's in use.
3. rest fields are needed for service needs

When the downloadingWorker receives this entity, it increases the counter `activeQueriers` and creates blockQuerier wrapper that has `Close` function that will decrease the counter once blockQuerier is not needed anymore. 

When the cache entry is being removed from the cache, the cache calls the function `removeDirectoryAsync` which asynchronously removes the block's folder. This function checks every `Xms` if there are still active block queriers and once `activeQueriers` count is `0` the folder will be removed. 
Also, there is a timeout in this function, and once the timeout is reached, the folder will be force removed.

**Special notes for your reviewer**:
* If the cache is disabled, then the blocks will be downloaded each time the block is requested.
*  If the cache is used, the folder will be deleted by embeddedCache when it reaches memory size limit or items count limit. Otherwise, the block's folder will be deleted when `Close` function is called.

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [ ] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go [hunk 1]
```diff
@@ -236,3 +298,108 @@ func extractArchive(archivePath string, workingDirectoryPath string) error {
...
+		},
+	}
+}
+
+func newBlockQuerierFromFS(blockDirectory string) *closableBlockQuerier {
+	return &closableBlockQuerier{
+		BlockQuerier: createBlockQuerier(blockDirectory),
+		Close: func() error {
+			return deleteFolder(blockDirectory)
+		},
```
[the remark this answers is not in the document]
Comment (chaudum) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader.go:323]: Ok, I think I understand now why you implemented it like this.

**Review by chaudum — COMMENTED** [review]
Overall looks quite good already. I think I would prefer a noop-cache implementation instead of the nil-check, because it would unify the code paths.

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go [hunk 2]
```diff
@@ -192,8 +254,11 @@ func (d *blockDownloader) extractBlock(block *Block, ts time.Time) (string, erro
 		return "", fmt.Errorf("error writing data to temp file: %w", err)
 	}
 	defer func() {
-		os.Remove(archivePath)
-		// todo log err
+		err = os.Remove(archivePath)
+		if err == nil {
+			return
+		}
+		level.Error(logger).Log("msg", "error removing temp archive file", "err", err)
```
[the remark this answers is not in the document]
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader.go:261]: ;) do not know how it happened ;) fixed

File: pkg/storage/stores/shipper/bloomshipper/block_downloader.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader.go:323]: If we do not use the cache, then we can not control the total size of the files on FS, so, it's better to remove it once we complete the processing. 
in our case, we will always use the cache, but some OSS users might have it disabled (default state)

File: pkg/storage/stores/shipper/bloomshipper/block_downloader_test.go [hunk 3]
```diff
@@ -67,37 +71,285 @@ func Test_blockDownloader_downloadBlocks(t *testing.T) {
...
+			require.NoError(t, multiError.Err())
+			require.Equal(t, uint32(10), blocksDownloadedCount.Load())
+			require.Equal(t, testData.expectedTotalGetBlocksCalls, blockClient.getBlockCalls)
+		})
+	}
+}
+
+func Test_cachedBlock(t *testing.T) {
+	tests := map[string]struct {
+		releaseQierier                   bool
```
[the remark this answers is not in the document]
Comment (vlad-diachenko) [reply pkg/storage/stores/shipper/bloomshipper/block_downloader_test.go:233]: thanks. fixed

File: docs/sources/configure/_index.md [hunk 4]
```diff
@@ -2281,6 +2281,29 @@ bloom_shipper:
...
     [max_tasks_enqueued_per_tenant: <int> | default = 10000]
+
+  blocks_cache:
+    # Whether embedded cache is enabled.
+    # CLI flag: -blocks-cache.enabled
+    [enabled: <boolean> | default = false]
+
+    # Maximum memory size of the cache in MB.
+    # CLI flag: -blocks-cache.max-size-mb
+    [max_size_mb: <int> | default = 100]
```
Comment (chaudum) [inline docs/sources/configure/_index.md:2292]: Unfortunately we cannot use a better default, since the config is used in multiple places :(

**Review by chaudum — APPROVED** [review]

File: docs/sources/configure/_index.md (hunk 4 above)
Comment (vlad-diachenko) [reply docs/sources/configure/_index.md:2297]: we can, but I do not know what is suitable value here... it depends on the customer, and how much data they have, and it's important if they have enough disk space...  let's come back to this question maybe later ...

**Pull request merged**